File: Musixmatch_lyrics_API_App/app/src/main/java/alexejantonov/com/musixmatch_lyrics_api_app/utils/DataMergeUtil.java
package alexejantonov.com.musixmatch_lyrics_api_app.utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.artist.Artist;
import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.track.Track;
import alexejantonov.com.musixmatch_lyrics_api_app.ui.Base.BaseData;

public class DataMergeUtil {

	public DataMergeUtil() {
	}

	public static List<BaseData> listsMerge(List<Artist> artists, List<Track> tracks) {
		List<BaseData> data = new ArrayList<>();

		for (int i = 0; i < artists.size(); i++) {
			data.add(artists.get(i));
			for (int j = 0; j < tracks.size(); j++) {
				if (artists.get(i).getArtistId() == tracks.get(j).getArtistId()) {
					data.add(tracks.get(j));
				}
			}
		}
		return data;
	}

	public static List<BaseData> searchListsMerge(List<Artist> artists, List<Track> tracks, List<BaseData> data) {
		List<BaseData> mergeData = new ArrayList<>();
		Map<Artist, List<Track>> dataMap = new HashMap<>();

		for (int i = 0; i < data.size(); i++) {
			if (data.get(i) instanceof Artist) {
				for (int j = 0; j < tracks.size(); j++) {
					if (data.get(i).getArtistId() == tracks.get(j).getArtistId()) {
						if (dataMap.containsKey(data.get(i))) {
							dataMap.get(data.get(i)).add(tracks.get(j));
						} else {
							List<Track> mapBucketTracks = new ArrayList<>();
							mapBucketTracks.add(tracks.get(j));
							dataMap.put((Artist) data.get(i), mapBucketTracks);
						}
					}
				}
			} else if (data.get(i) instanceof Track) {
				for (int j = 0; j < artists.size(); j++) {
					if (data.get(i).getArtistId() == artists.get(j).getArtistId()) {
						if (dataMap.containsKey(artists.get(j))) {
							dataMap.get(artists.get(j)).add((Track) data.get(i));
							break;
						} else {
							List<Track> mapBucketTrack = new ArrayList<>();
							mapBucketTrack.add((Track) data.get(i));
							dataMap.put(artists.get(j), mapBucketTrack);
							break;
						}
					}
				}
			}
		}

		for (Artist artist : dataMap.keySet()) {
			mergeData.add(artist);
			mergeData.addAll(dataMap.get(artist));
		}

		return mergeData;
	}
}

File: Musixmatch_lyrics_API_App/app/src/main/java/alexejantonov/com/musixmatch_lyrics_api_app/utils/DataMergeUtil.kt
package alexejantonov.com.musixmatch_lyrics_api_app.utils

import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.artist.Artist
import alexejantonov.com.musixmatch_lyrics_api_app.api.entities.track.Track
import alexejantonov.com.musixmatch_lyrics_api_app.ui.Base.BaseData

object DataMergeUtil {

  fun listsMerge(artists: List<Artist>, tracks: List<Track>): List<BaseData> {
    val data = mutableListOf<BaseData>()

    for (artist in artists) {
      data.add(artist)
      for (track in tracks) {
        if (artist.artistId == track.artistId) {
          data.add(track)
        }
      }
    }
    return data
  }

  fun searchListsMerge(artists: List<Artist>, tracks: MutableList<Track>, query: String): List<BaseData> {
    val mergeData = mutableListOf<BaseData>()
    val tracksClone = mutableListOf<Track>().apply { addAll(tracks) }

    for (artist in artists) {
      if (artist.artistName.toLowerCase().contains(query.toLowerCase())) {
        mergeData.add(artist)
        for (track in tracks) {
          if (track.artistId == artist.artistId) {
            mergeData.add(track)
            tracksClone.remove(track)
          }
        }
      }
    }

    for (track in tracksClone) {
      if (track.trackName.toLowerCase().contains(query.toLowerCase())) {
        for (artist in artists) {
          if (artist.artistId == track.artistId) {
            if (mergeData.contains(artist).not()) {
              mergeData.add(artist)
            }
          }
        }
        mergeData.add(track)
      }
    }
    return mergeData
  }
}
